src/main/java/net/neoforged/neoforge/transfer/ResourceHandlerUtil.java

```diff
@@ -0,0 +1,565 @@
+            IResourceHandler<R> handler,
+            Predicate<R> filter,
+            int amount,
+            R defaultResource,
+            @Nullable TransactionContext transaction,
+            IStackFactory<R, S> stackFactory) {
+        try (Transaction subTransaction = TransactionManager.open(transaction)) {
+            int size = handler.size();
+            int handled = 0;
+            R resourceTarget = defaultResource;
+            for (int index = 0; index < size; index++) {
+                R resource = handler.getResource(index);
+                if (resource.isEmpty()) continue;
+                if (!filter.test(resource)) continue;
+                if (resourceTarget.isEmpty())
+                    resourceTarget = resource;
+                else if (!resourceTarget.equals(resource)) continue;
+
+                handled += handler.extract(resource, amount - handled, subTransaction);
+                if (handled == amount) {
+                    subTransaction.commit();
+                    return stackFactory.create(resource, handled);
+                }
+            }
+
+            subTransaction.commit();
+            if (handled == 0) return stackFactory.create(defaultResource, 0);
+            return stackFactory.create(resourceTarget, handled);
+        }
+    }
+
+    /**
+     * Extracts the first resource from an {@link IResourceHandler} that is not empty.
+     *
+     * @param <R>          the type of resource handled by the handler.
+     * @param <S>          The type of stack returned by the handler.
+     * @param index        The index that is being checked in the handler.
+     * @param handler      the {@link IResourceHandler} to extract the resource from.
+     * @param amount       the desired amount of the resource to extract.
+     * @param transaction  The transaction context for a given insertion.
+     *                     Passing in {@code null} will essentially be the same as doing `execute`,
+     *                     whereas passing in a closeable context allows you to choose if it should be committed.
+     * @param stackFactory A factory the given a resource of type {@code <R>} and an amount, a stack of type {@code <S>} can be created. The return is expected to be non-null and properly be the instanced empty value for a given resource.
+     * @return a stack of type {@code <S>} typically in the form of an ResourceStack or as an example an ItemStack based on the factory provided
+     */
+    public static <R extends IResource, S> S extractIndexFiltered(
+            IResourceHandler<R> handler,
+            Predicate<R> filter,
+            int index,
+            int amount,
+            R defaultResource,
+            @Nullable TransactionContext transaction,
+            IStackFactory<R, S> stackFactory) {
+        try (Transaction subTransaction = TransactionManager.open(transaction)) {
+            R resource = handler.getResource(index);
```

Comment:
This logic that gets the resource for the handler and tests it, and then returns default if it doesn't match can happen before we open a transaction in the first place.